src/Init/Data/String/Slice.lean

```diff
@@ -0,0 +1,1352 @@
+  .defaultImplementation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorLoop (PosIterator s) m n :=
+  .defaultImplementation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorLoopPartial (PosIterator s) m n :=
+  .defaultImplementation
+
+docs_to_verso positions
+
+end PosIterator
+
+structure RevPosIterator (s : Slice) where
+  currPos : s.Pos
+  deriving Inhabited
+
+set_option doc.verso false
+/--
+Creates and iterator over all valid positions within {name}`s`, starting from the last valid
+position and iterating towards the first one.
+
+Examples
+ * {lean}`("abc".toSlice.revPositions.map (·.get!) |>.toList) = ['c', 'b', 'a']`
+ * {lean}`("abc".toSlice.revPositions.map (·.offset.byteIdx) |>.toList) = [2, 1, 0]`
+ * {lean}`("ab∀c".toSlice.revPositions.map (·.get!) |>.toList) = ['c', '∀', 'b', 'a']`
+ * {lean}`("ab∀c".toSlice.revPositions.map (·.offset.byteIdx) |>.toList) = [5, 2, 1, 0]`
+-/
+def revPositions (s : Slice) : Std.Iter (α := RevPosIterator s) s.Pos :=
+  { internalState := { currPos := s.endPos }}
+
+set_option doc.verso true
+
+namespace RevPosIterator
+
+instance [Pure m] : Std.Iterators.Iterator (RevPosIterator s) m s.Pos where
+  IsPlausibleStep it
+    | .yield it' out =>
+      ∃ h : it.internalState.currPos ≠ s.startPos,
+        it'.internalState.currPos = it.internalState.currPos.prev h ∧
+        it.internalState.currPos.prev h = out
+    | .skip _ => False
+    | .done => it.internalState.currPos = s.startPos
+  step := fun ⟨⟨currPos⟩⟩ =>
+    if h : currPos = s.startPos then
+      pure ⟨.done, by simp [h]⟩
+    else
+      let prevPos := currPos.prev h
+      pure ⟨.yield ⟨⟨prevPos⟩⟩ prevPos, by simp [h, prevPos]⟩
+
+private def finitenessRelation [Pure m] :
+    Std.Iterators.FinitenessRelation (RevPosIterator s) m where
+  rel := InvImage WellFoundedRelation.rel
+      (fun it => it.internalState.currPos.offset.byteIdx)
+  wf := InvImage.wf _ WellFoundedRelation.wf
+  subrelation {it it'} h := by
+    simp_wf
+    obtain ⟨step, h, h'⟩ := h
+    cases step
+    · cases h
+      obtain ⟨h1, h2, _⟩ := h'
+      have h3 := Pos.offset_prev_lt_offset (h := h1)
+      simp [Pos.ext_iff, String.Pos.ext_iff] at h2 h3
+      omega
+    · cases h'
+    · cases h
+
+@[no_expose]
+instance [Pure m] : Std.Iterators.Finite (RevPosIterator s) m :=
+  .of_finitenessRelation finitenessRelation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorCollect (RevPosIterator s) m n :=
+  .defaultImplementation
+
+instance [Monad m] [Monad n] :
+    Std.Iterators.IteratorCollectPartial (RevPosIterator s) m n :=
+  .defaultImplementation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorLoop (RevPosIterator s) m n :=
+  .defaultImplementation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorLoopPartial (RevPosIterator s) m n :=
+  .defaultImplementation
+
+docs_to_verso revPositions
+
+end RevPosIterator
+
+structure ByteIterator where
```

Comment:
Looking forward to deprecating this again when we can convert `String.Slice` to `ByteSlice` :) But doing so efficiently will require the runtime representation of `String` to contain an actual `ByteArray`.